Sources/SwiftParser/Specifiers.swift

```diff
@@ -415,7 +415,108 @@ extension RawAccessorEffectSpecifiersSyntax: RawEffectSpecifiersTrait {
       }
     }
   }
+}
+
+// This conformance exists solely to parse misplaced specifiers in parseFunctionReturnClause()
+extension RawDeinitEffectSpecifiersSyntax: RawEffectSpecifiersTrait {
+  enum MisspelledAsyncSpecifiers: TokenSpecSet {
+    case await
+    case reasync
+
+    init?(lexeme: Lexer.Lexeme) {
+      switch PrepareForKeywordMatch(lexeme) {
+      case TokenSpec(.await, allowAtStartOfLine: false): self = .await
+      case TokenSpec(.reasync): self = .reasync
+      default: return nil
+      }
+    }
+
+    var spec: TokenSpec {
+      switch self {
+      case .await: return TokenSpec(.await, allowAtStartOfLine: false)
+      case .reasync: return .keyword(.reasync)
+      }
+    }
+  }
+
+  enum CorrectAsyncTokenKinds: TokenSpecSet {
+    case async
+
+    init?(lexeme: Lexer.Lexeme) {
+      switch PrepareForKeywordMatch(lexeme) {
+      case TokenSpec(.async): self = .async
+      default: return nil
+      }
+    }
+
+    var spec: TokenSpec {
+      switch self {
+      case .async: return .keyword(.async)
+      }
+    }
+  }
+
+  enum MisspelledThrowsTokenKinds: TokenSpecSet {
+    case `rethrows`
+    case `try`
+    case `throw`
+    case `throws`
+
+    init?(lexeme: Lexer.Lexeme) {
+      switch PrepareForKeywordMatch(lexeme) {
+      case TokenSpec(.rethrows): self = .rethrows
+      case TokenSpec(.try, allowAtStartOfLine: false): self = .try
+      case TokenSpec(.throw, allowAtStartOfLine: false): self = .throw
+      case TokenSpec(.throws): self = .throws
+      default: return nil
+      }
+    }
+
+    var spec: TokenSpec {
+      switch self {
+      case .rethrows: return .keyword(.rethrows)
+      case .try: return TokenSpec(.try, allowAtStartOfLine: false)
+      case .throw: return TokenSpec(.throw, allowAtStartOfLine: false)
+      case .throws: return .keyword(.throws)
+      }
+    }
+  }
+
+  enum CorrectThrowsTokenKinds: TokenSpecSet {
+    // Uninhabited
```

Comment:
We don’t have any precedence for empty `TokenSpecSet` and `canRecoveryTo` has a precondition that the spec set is not empty. 

Could you change 

https://github.com/apple/swift-syntax/blob/8a759449c646c72b9392b151236204a5fad03831/Sources/SwiftParser/Recovery.swift#L133

to 

```swift
if specSet.allCases.isEmpty {
  return nil
}
```

Reply:
Done

Reply:
Actually, in the single use case where this conformance is used, `canRecoverTo()` is not used, only `at()` which seems to be perfectly fine with empty spec sets. So, I guess changes in `canRecoverTo()` are not needed. But I don't mind to do them just in case. @ahoppen, WDYT?

Reply:
I think allowing empty `TokenSpecSet` makes total sense in `canRecoverTo` and I‘d prefer to be safe here instead of discovering a stupid crash about it later if we for some reason start to use `canRecoverTo` in the future.